app/src/org/commcare/tasks/UnzipTask.java

```diff
@@ -1,109 +1,137 @@
     public static final int UNZIP_TASK_ID = 7212435;
 
-    public UnzipTask() {
+    protected UnzipTask() {
         this.taskId = UNZIP_TASK_ID;
-
         TAG = UnzipTask.class.getSimpleName();
     }
 
     @Override
     protected Integer doTaskBackground(String... params) {
-        File archive = new File(params[0]);
-        File destination = new File(params[1]);
+        if (params[0].startsWith("content://")) {
+            return unZipFromContentUri(params[0], params[1]);
+        } else {
+            return unZipFromFilePath(params[0], params[1]);
+        }
+    }
 
-        Log.d(TAG, "Unzipping archive '" + archive + "' to  '" + destination + "'");
+    private Integer unZipFromFilePath(String filePath, String destinationPath) {
+        File archive = new File(filePath);
+        ZipInputStream zis;
+        try {
+            zis = new ZipInputStream(new FileInputStream(archive));
+        } catch (FileNotFoundException e) {
+            publishProgress(getInvalidZipFileErrorMessage());
+            return -1;
+        }
+        return unZipFromStream(zis, destinationPath);
+    }
 
-        int count = 0;
-        ZipFile zipfile;
-        //From stackexchange
+    private Integer unZipFromContentUri(String uriString, String destinationPath) {
+        // we have a contenturi, use it to get the InputStream
+        Uri fileUri = Uri.parse(uriString);
+        ZipInputStream zis;
         try {
-            zipfile = new ZipFile(archive);
-        } catch (IOException ioe) {
-            publishProgress("Could not find target file for unzipping.");
+            InputStream is = CommCareApplication.instance().getContentResolver().openInputStream(fileUri);
+            zis = new ZipInputStream(is);
+        } catch (FileNotFoundException e) {
+            publishProgress(getInvalidZipFileErrorMessage());
             return -1;
         }
-        for (Enumeration e = zipfile.entries(); e.hasMoreElements(); ) {
-            Localization.get("mult.install.progress", new String[]{String.valueOf(count)});
-            count++;
-            ZipEntry entry = (ZipEntry)e.nextElement();
+        return unZipFromStream(zis, destinationPath);
+    }
 
-            if (entry.isDirectory()) {
-                FileUtil.createFolder(new File(destination, entry.getName()).toString());
-                //If it's a directory we can move on to the next one
-                continue;
-            }
+    private Integer unZipFromStream(ZipInputStream zis, String destinationPath) {
+        File destination = new File(destinationPath);
+        int count = 0;
+        ZipEntry entry = null;
+        try {
+            while ((entry = zis.getNextEntry()) != null) {
+                publishProgress(Localization.get("mult.install.progress", new String[]{String.valueOf(count)}));
+                count++;
 
-            File outputFile = new File(destination, entry.getName());

```

Comment:
While unzipping, Instead of operating on file, we are now operating on a stream now